src/charm.py

```diff
@@ -6,20 +6,355 @@
+        - `juju_unit`
+    """
+
+    def __init__(self, template: str, key: str, *args):
+        msg = textwrap.dedent(
+            f"""Unable to render the template {template!r}: {key!r} unknown.
+                - `juju_model`
+                - `juju_application`
+                - `juju_unit`"""
+        )
+        super().__init__(msg, *args)
+
+
+@dataclass
+class RouteConfig:
+    """Route configuration."""
+
+    root_url: str
+    rule: str
+    id_: str
+
+
+@dataclass
+class _RouteConfig:
+    root_url: str
+    rule: Optional[str] = None
+
+    @property
+    def is_valid(self) -> bool:
+        def _check_var(obj: str, name: str) -> bool:
+            error = None
+            if not _is_not_empty(obj):
+                error = (
+                    f"`{name}` not configured; run `juju config oathkeeper-configurator"
+                    f"{name}=<{name.upper()}>"
+                )
+
+            elif obj != (stripped := obj.strip()):
+                error = (
+                    f"{name} {obj!r} starts or ends with whitespace;" f"it should be {stripped!r}."
+                )
+
+            if error:
+                logger.error(error)
+            return not error
+
+        if self.root_url:
+            try:
+                # try rendering with dummy values; it should succeed.
+                self.render(model_name="foo", unit_name="bar", app_name="baz")
+            except (TemplateKeyError, RuleDerivationError) as e:
+                logger.info("Failed to render a template with dummy values")
+                logger.error(e)
+                return False
+
+        if not self.rule:
+            # the rule can be guessed from the root_url
+            return _check_var(self.root_url, "root_url")
+
+        valid = starmap(_check_var, ((self.rule, "rule"), (self.root_url, "root_url")))
+        return all(valid)
+
+    def render(self, model_name: str, unit_name: str, app_name: str) -> "RouteConfig":
+        """Fills in the blanks in the templates."""
+
+        def _render(obj: str) -> Optional[str]:
+            # StrictUndefined will raise an exception if some undefined
+            # variables are left unrendered in the template
+            template = jinja2.Template(obj, undefined=jinja2.StrictUndefined)
+            try:
+                return template.render(
+                    juju_model=model_name, juju_application=app_name, juju_unit=unit_name
+                )
+            except jinja2.UndefinedError as e:
+                undefined_key = e.message.split()[0].strip(r"'")
+                raise TemplateKeyError(obj, undefined_key) from e
+
+        url = _render(self.root_url)
+        if not self.rule:
+            rule = self.generate_rule_from_url(url)
+        else:
+            rule = _render(self.rule)
+
+        # an easily recognizable id for the traefik services
+        id_ = "-".join((unit_name, model_name))
+        return RouteConfig(rule=rule, root_url=url, id_=id_)
+
+    @staticmethod
```

Comment:
Not sure why this is a `staticmethod` and not a function.

Reply:
it's thematically related to the class even though it doesn't need access to it, so I would think it's a staticmethod